Feature/dem 805/implement sampling rate float parameter

qSaevar opened the pull request:
Changes proposed in this pull request:
- Set and Get the sampling rate of the ZIUHFLI using float. 


@WilliamHPNielsen @jenshnielsen @astafan8

astafan8 commented on qcodes/instrument_drivers/ZI/ZIUHFLI.py:
```diff
@@ -1330,6 +1330,20 @@ def __init__(self, name: str, device_ID: str, **kwargs) -> None:
+                           set_cmd=partial(self._set_samplingrate_as_float,
+                                           self._samplingrate_codes),
+                           unit='Hz',
+                           get_cmd=partial(self._get_samplingrate_as_float),
```
no need for partial here

qSaevar replied:
✔️

astafan8 commented on qcodes/instrument_drivers/ZI/ZIUHFLI.py:
```diff
@@ -2168,6 +2182,23 @@ def getduration():
+        nearest_frequency = min(samplingrate_codes.keys(), key=lambda k: abs(
+            frequency - ZIUHFLI._convert_to_float(k)))
+
+        self.scope_samplingrate(nearest_frequency)
```
smth like `self.scope_samplingrate_float._save_val(ZIUHFLI._convert_to_float(nearest_frequency))` is missing here, right, @QCoDeS/core ?

qSaevar replied:
right?

astafan8 replied:
@QCoDeS/core ? :)

WilliamHPNielsen replied:
See review below.

astafan8 commented on qcodes/instrument_drivers/ZI/ZIUHFLI.py:
```diff
@@ -1330,6 +1330,20 @@ def __init__(self, name: str, device_ID: str, **kwargs) -> None:
+                                           self._samplingrate_codes),
+                           unit='Hz',
+                           get_cmd=partial(self._get_samplingrate_as_float),
+                           docstring=""" Set the scope sampling rate with float.
```
i personally think the word "float" is abused in this PR. smth like "numeric" or "number" seems to be better. but i leave this to the author.

qSaevar replied:
✔️

qSaevar commented:
@astafan8 I addressed the your comments except the one with `self.scope_samplingrate_float._save_val(ZIUHFLI._convert_to_float(nearest_frequency))` as it was left with a question mark. Should I do it?

qSaevar commented:
@WilliamHPNielsen I very well see your point and I'm not going to try and convince you. This behavior of setting the nearest point is how it works in the m4i driver, so there was a request for it in the uhfli driver as well.
But having the option of excepting and returning the sampling rate as float would still be very handy to have in many cases. So what if I make the "nearest" frequency a static method that finds the nearest allowed frequency, given some floating point number, that can in turn be used to set the sampling frequency? Something like:
```
sampling_rate = ZIUHFLI.round_to_nearest_sampling_frequency(desired_sampling_rate)
uhfli.scope_samplingrate_float(sampling_rate)
```
?

WilliamHPNielsen commented:
@qSaevar I think what you propose with
```python
sampling_rate = ZIUHFLI.round_to_nearest_sampling_frequency(desired_sampling_rate)
uhfli.scope_samplingrate_float(sampling_rate)
```
is a nice and clean solution. 

Then the `scope_samplingrate_float` parameter can have an `Enum` validator only accepting the correct values, and the `round_to_nearest_sampling_frequency` method should then take care not to introduce floating-point errors, but that should not be too hard to implement.

qSaevar commented:
@WilliamHPNielsen I updated the PR so now the `scope_samplingrate_float` parameter only accepts sampling rates that are allowed by the device. There is also a helping method that returns the closest sampling rate from some desired value and it is no longer linear as before.

WilliamHPNielsen commented:
Looks terrific!
